client/pruning/main.go

```diff
@@ -0,0 +1,118 @@
+
+	"github.com/cosmos/cosmos-sdk/client/flags"
+	pruningtypes "github.com/cosmos/cosmos-sdk/pruning/types"
+	"github.com/cosmos/cosmos-sdk/server"
+	servertypes "github.com/cosmos/cosmos-sdk/server/types"
+	"github.com/cosmos/cosmos-sdk/store/rootmulti"
+	"github.com/tendermint/tendermint/libs/log"
+	dbm "github.com/tendermint/tm-db"
+)
+
+const FlagAppDBBackend = "app-db-backend"
+
+// PruningCmd prunes the sdk root multi store history versions based on the pruning options
+// specified by command flags.
+func PruningCmd(appCreator servertypes.AppCreator) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "prune",
+		Short: "Prune app history states by keeping the recent heights and deleting old heights",
+		Long: `Prune app history states by keeping the recent heights and deleting old heights.
+		The pruning option is provided via the '--pruning' flag or alternatively with '--pruning-keep-recent'
+		
+		For '--pruning' the options are as follows:
+		
+		default: the last 362880 states are kept
+		nothing: all historic states will be saved, nothing will be deleted (i.e. archiving node)
+		everything: 2 latest states will be kept
+		custom: allow pruning options to be manually specified through 'pruning-keep-recent'.
+
+		besides pruning options, database home directory and database backend type should also be specified via flags
+		'--home' and '--app-db-backend'.
+		valid app-db-backend type includes 'goleveldb', 'cleveldb', 'rocksdb', 'boltdb', and 'badgerdb'.
+		`,
+		Example: "prune --home './' --app-db-backend 'goleveldb' --pruning 'custom' --pruning-keep-recent 100",
+		RunE: func(cmd *cobra.Command, _ []string) error {
+			vp := viper.New()
+
+			// Bind flags to the Context's Viper so we can get pruning options.
+			if err := vp.BindPFlags(cmd.Flags()); err != nil {
+				return err
+			}
+			pruningOptions, err := server.GetPruningOptionsFromFlags(vp)
+			if err != nil {
+				return err
+			}
+			fmt.Printf("get pruning options from command flags, strategy: %v, keep-recent: %v\n",
+				pruningOptions.Strategy,
+				pruningOptions.KeepRecent,
+			)
+
+			home := vp.GetString(flags.FlagHome)
+			db, err := openDB(home, server.GetAppDBBackend(vp))
+			if err != nil {
+				return err
+			}
+
+			logger := log.NewTMLogger(log.NewSyncWriter(os.Stdout))
+			app := appCreator(logger, db, nil, vp)
+			cms := app.CommitMultiStore()
+
+			rootMultiStore, ok := cms.(*rootmulti.Store)
+			if !ok {
+				return fmt.Errorf("currently only support the pruning of rootmulti.Store type")
+			}
+			latestHeight := rootmulti.GetLatestVersion(db)
+			// valid heights should be greater than 0.
+			if latestHeight <= 0 {
+				return fmt.Errorf("the database has no valid heights to prune, the latest height: %v", latestHeight)
+			}
+
+			var pruningHeights []int64
+			for height := int64(1); height < latestHeight; height++ {
+				if height < latestHeight-int64(pruningOptions.KeepRecent) {
+					pruningHeights = append(pruningHeights, height)
```

Comment:
Should it check the other prunning flags here?

Reply:
`pruning-interval` is only used to decide at which height the pruning action would happen.
but actually for every height whether it is to-be-pruned only depends on the `keep-recent` and the `snapshot-options`, so almost every height older than `keep-recent` will be moved to pruneHeights after the snapshot completes.
refer to: https://github.com/cosmos/cosmos-sdk/blob/main/pruning/manager.go#L102-L152